src/Chainweb/Cut/Create.hs

```diff
@@ -0,0 +1,442 @@
+import Chainweb.Difficulty
+import Chainweb.Payload
+import Chainweb.Time
+import Chainweb.Utils
+import Chainweb.Version
+import Chainweb.Version.Utils
+
+-- -------------------------------------------------------------------------- --
+-- Adjacent Parent Hashes
+
+-- | Witnesses that a cut can be extended for the respective header.
+--
+data CutExtension = CutExtension
+    { _cutExtensionCut' :: !Cut
+        -- This is overly restrictive, since the same cut extension can be
+        -- valid for more than one cut. It's fine for now.
+    , _cutExtensionParent' :: !ParentHeader
+    , _cutExtensionAdjacentHashes' :: !BlockHashRecord
+    }
+
+makeLenses ''CutExtension
+
+_cutExtensionCut :: CutExtension -> Cut
+_cutExtensionCut = _cutExtensionCut'
+
+cutExtensionCut :: Lens' CutExtension Cut
+cutExtensionCut = cutExtensionCut'
+
+_cutExtensionParent :: CutExtension -> ParentHeader
+_cutExtensionParent = _cutExtensionParent'
+
+cutExtensionParent :: Lens' CutExtension ParentHeader
+cutExtensionParent = cutExtensionParent'
+
+_cutExtensionAdjacentHashes :: CutExtension -> BlockHashRecord
+_cutExtensionAdjacentHashes = _cutExtensionAdjacentHashes'
+
+cutExtensionAdjacentHashes :: Lens' CutExtension BlockHashRecord
+cutExtensionAdjacentHashes = cutExtensionAdjacentHashes'
+
+instance HasChainId CutExtension where
+    _chainId = _chainId . _cutExtensionParent
+    {-# INLINE _chainId #-}
+
+instance HasChainwebVersion CutExtension where
+    _chainwebVersion = _chainwebVersion . _cutExtensionCut
+    {-# INLINE _chainwebVersion #-}
+
+-- | Witness that a cut can be extended for the given chain by trying to
+-- assemble the adjacent hashes for a new work header.
+--
+-- If there is a graph change the adajacent parents on the new chains are the
+-- genesis parents.
+--
+-- Graph changes add some complexity to this function. There are three cases:
+--
+-- 1. This is a transition cut, i.e. the cut contains block headers with
+--    different graphs. In this case we wait until all chains transitions to the
+--    new graph before add the genesis blocks for the new chains and move ahead.
+--
+-- 2. This is a transition step (parentGraph /= targetGraph). Only do the step
+--    when new /and/ old adjacents headers are available. The old dependencies
+--    are not used but are needed to guarantee that the cut can extended on
+--    other chains.
+--
+-- 3. This is a "normal" transition.
+--
+-- TODO: it is important that the semantics of this function corresponds to the
+-- respective validation in the module "Chainweb.Cut", in particular
+-- 'isMonotonicCutExtension'. It must not happen, that a cut passes validation
+-- that can't be further extended.
+--
+getCutExtension
+    :: HasCallStack
+    => HasChainId cid
+    => Cut
+        -- ^ the cut which is to be extended
+    -> cid
+        -- ^ the header onto which the new block is created. It is expected
+        -- that this header is contained in the cut.
+    -> Maybe CutExtension
```

Comment:
Consider making this an `Either XXX CutExtension` so we can see why it failed, given there are multiple failure modes.

Reply:
Changing the result to `Either` makes sense, although mostly for debugging and logging purposes.

This function will be refactored in an upcoming PR. Most likely it will be merged with the respective function in the Cut module. For now it returns `Maybe` because that's what the existing Coordinator was expecting.